File: package/go/src/markdown_to_html.go
package src

import (
	"regexp"
	"strings"
)

// MarkdownToHTML converts a markdown format string to HTML format string.
//
// Supported syntax:
//   - Headings: Lines starting with '#' are converted to <h1>~<h6> tags based on the number of '#'
//   - Strong: Text surrounded by ** is converted to <strong> tags
//   - Italic: Text surrounded by * is converted to <em> tags
//   - Code: Text surrounded by ` is converted to <code> tags (no conversion inside code)
//   - Paragraphs: Blocks separated by empty lines are converted to <p> tags
//
// Parameters:
//   - md: markdown format string
//
// Returns:
//   - HTML format string
func MarkdownToHTML(md string) string {
	// Split input text into lines
	lines := strings.Split(md, "\n")
	var blocks [][]string
	var currentBlock []string

	// Split into blocks separated by empty lines (ignore empty blocks)
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			if len(currentBlock) > 0 {
				blocks = append(blocks, currentBlock)
				currentBlock = []string{}
			}
		} else {
			currentBlock = append(currentBlock, line)
		}
	}
	if len(currentBlock) > 0 {
		blocks = append(blocks, currentBlock)
	}

	var htmlLines []string
	headingPattern := regexp.MustCompile(`^(#{1,6})\s+(.*)$`)

	for _, block := range blocks {
		// If block has only one line and starts with '#', treat as heading
		if len(block) == 1 {
			matches := headingPattern.FindStringSubmatch(block[0])
			if matches != nil {
				level := len(matches[1])
				content := matches[2]
				content = replaceInline(content)
				htmlLines = append(htmlLines, "<h"+string(rune(level+'0'))+">"+content+"</h"+string(rune(level+'0'))+">")
				continue
			}
		}

		// Multi-line blocks or single lines that are not headings are treated as paragraphs
		var paragraphParts []string
		for _, line := range block {
			paragraphParts = append(paragraphParts, strings.TrimSpace(line))
		}
		paragraph := strings.Join(paragraphParts, " ")
		paragraph = replaceInline(paragraph)
		htmlLines = append(htmlLines, "<p>"+paragraph+"</p>")
	}

	return strings.Join(htmlLines, "\n")
}

// replaceInline converts inline markdown syntax (strong, italic, code) to HTML tags.
// Code sections are protected from other conversions.
//
// Parameters:
//   - text: text containing inline syntax
//
// Returns:
//   - text with inline syntax replaced with HTML tags
func replaceInline(text string) string {
	// Split inline code temporarily using regex
	codePattern := regexp.MustCompile("`[^`]+`")
	parts := codePattern.Split(text, -1)
	codeParts := codePattern.FindAllString(text, -1)

	// Apply strong and italic conversion to non-code parts only
	strongPattern := regexp.MustCompile(`\*\*([^*]+)\*\*`)
	italicPattern := regexp.MustCompile(`\*([^*]+)\*`)

	for i, part := range parts {
		// **...** -> <strong>...</strong>
		part = strongPattern.ReplaceAllString(part, "<strong>$1</strong>")
		// *...* -> <em>...</em>
		part = italicPattern.ReplaceAllString(part, "<em>$1</em>")
		parts[i] = part
	}

	// Reassemble with code parts converted to <code> tags
	var result strings.Builder
	codeIndex := 0
	for i, part := range parts {
		result.WriteString(part)
		if i < len(parts)-1 && codeIndex < len(codeParts) {
			// Convert `code` to <code>code</code>
			codeContent := codeParts[codeIndex]
			if strings.HasPrefix(codeContent, "`") && strings.HasSuffix(codeContent, "`") {
				innerCode := codeContent[1 : len(codeContent)-1]
				result.WriteString("<code>" + innerCode + "</code>")
			}
			codeIndex++
		}
	}

	return result.String()
}

File: package/typescript/src/markdownToHtml.ts
/**
 * マークダウン形式の文字列をHTML形式の文字列に変換する。
 *
 * サポートする記法:
 *   見出し: 行頭の '#' の個数に応じた <h1>〜<h6> タグで囲む
 *   強調: **で囲まれた部分を <strong> タグで囲む
 *   イタリック: *で囲まれた部分を <em> タグで囲む
 *   コード: `で囲まれた部分を <code> タグで囲む（コード内は変換を行わない）
 *   段落: 空行で区切られたブロックを <p> タグで囲む
 *
 * @param md - マークダウン形式の文字列
 * @returns HTML形式の文字列
 */
export function markdownToHtml(md: string): string {
  // 入力テキストを行ごとに分割
  const lines = md.split('\n');
  const blocks: string[][] = [];
  let currentBlock: string[] = [];

  // 空行を境にブロックに分割する（空ブロックは無視）
  for (const line of lines) {
    if (line.trim() === '') {
      if (currentBlock.length > 0) {
        blocks.push(currentBlock);
        currentBlock = [];
      }
    } else {
      currentBlock.push(line);
    }
  }
  if (currentBlock.length > 0) {
    blocks.push(currentBlock);
  }

  const htmlLines: string[] = [];
  for (const block of blocks) {
    // ブロックが1行のみかつ行頭が '#' で始まる場合、見出しと判断する
    if (block.length === 1) {
      const match = block[0].match(/^(#{1,6})\s+(.*)$/);
      if (match) {
        const level = match[1].length;
        const content = match[2];
        const processedContent = replaceInline(content);
        htmlLines.push(`<h${level}>${processedContent}</h${level}>`);
        continue;
      }
    }

    // 複数行のブロックまたは見出しでない1行は段落として扱う
    const paragraph = block.map(line => line.trim()).join(' ');
    const processedParagraph = replaceInline(paragraph);
    htmlLines.push(`<p>${processedParagraph}</p>`);
  }

  return htmlLines.join('\n');
}

/**
 * テキスト中のインラインマークダウン記法（強調、イタリック、コード）をHTMLタグに変換する。
 * コードで囲まれた部分は保護し、他の変換の影響を受けないようにします。
 *
 * @param text - インライン記法を含むテキスト
 * @returns インライン記法をHTMLタグに置換したテキスト
 */
function replaceInline(text: string): string {
  // inline code を一時的に分離するために split を使用
  const parts = text.split(/(`[^`]+`)/);

  // parts のうち、インラインコードでない部分だけに強調やイタリックの変換を適用
  for (let i = 0; i < parts.length; i++) {
    const part = parts[i];
    if (!(part.startsWith('`') && part.endsWith('`'))) {
      // **...** -> <strong>...</strong>
      let processed = part.replace(/\*\*([^*]+)\*\*/g, '<strong>$1</strong>');
      // *...* -> <em>...</em>
      processed = processed.replace(/\*([^*]+)\*/g, '<em>$1</em>');
      parts[i] = processed;
    }
  }

  // インラインコード部分を <code> タグに変換（中身はそのまま）
  for (let i = 0; i < parts.length; i++) {
    const part = parts[i];
    if (part.startsWith('`') && part.endsWith('`')) {
      const codeContent = part.slice(1, -1);
      parts[i] = `<code>${codeContent}</code>`;
    }
  }

  return parts.join('');
}
